url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/episodes/managers/UpnpManager.java

```diff
@@ -0,0 +1,397 @@
+import org.w3c.dom.Document;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import stv6.episodes.BasicEpisode;
+import stv6.http.HttpRequestor;
+import stv6.http.request.Request;
+import stv6.http.request.Response;
+
+/**
+ * Theoretically generic UPNP manager, but specifically
+ *  designed to work with the PS3MediaServer UPNP server
+ *  
+ * @author dhleong
+ *
+ */
+public class UpnpManager extends AbstractManager {
+    
+    private static final List<Path> basePaths = new LinkedList<Path>();
+    static {
+        basePaths.add(new Path("0", ""));
+    }
+
+    private static final String CONTENT_PATH = "/upnp/control/content_directory";
+    private static final String BROWSE_ACTION = "urn:schemas-upnp-org:service:ContentDirectory:1#Search";
+    
+    private static final String CLASS_FOLDER = "object.container.storageFolder";
+    private static final String CLASS_VIDEO  = "object.item.videoItem";
+    
+    private static final String DEFAULT_PROFILE_FILENAME = "PMS.conf";
+    private static final String ENV_PROFILE_PATH = "PMS_PROFILE";
+    private static final String PROPERTY_PROFILE_PATH = "pms.profile.path";
+    /** TODO This is NOT necessarily constant, and we should add a setting */
+    private static final String PROFILE_DIRECTORY_NAME = "PMS";
+    
+    private static final Pattern RESULT_PATTERN = Pattern.compile("<Result>(.*)</Result>");
+    
+    private String broadIp;
+    private int upnpPort = -1;
+    private boolean isPms = false;
+    
+    private String localFoldersInfo;
+    private final List<Path> localFolders = new LinkedList<Path>();
+    
+    public UpnpManager(String broadcastingIp, int upnpPort) {
+        this.broadIp = broadcastingIp;
+        this.upnpPort = upnpPort;
+    }
+    
+    public UpnpManager(String broadcastIp, int upnpPort, String localFoldersInfo) {
+        this(broadcastIp, upnpPort);
+        
+        this.localFoldersInfo = localFoldersInfo;
+    }
+    
+    public UpnpManager(String broadcastIp, boolean isPms) {
+        this.broadIp = broadcastIp;
+        
+        if (!isPms) 
+            return;
+        
+        this.isPms = true;
+        
+        reload(); // make sure we can load everything
+    }
+
+    @Override
+    public void reload() {
+        if (isPms) {
+            
+            // we can use some hax to get the information from the PMS.conf file
+            File pmsConf = getPmsProfile();
+            if (pmsConf == null || !pmsConf.exists()) {
+                throw new RuntimeException("Couldn't find PMS configuration.");
+            }
+            
+            // reset these to defaults
+            localFoldersInfo = null; 
+            upnpPort = -1;
+            
+            // load port and local folders
+            try {
+                BufferedReader r = new BufferedReader(new FileReader(pmsConf));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *r*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at lines 126, 130 evaluate to true. Ensure *r* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).